routes/profile.js

```diff
@@ -52,54 +52,89 @@ async function getProfileData(profileName, geoids, compare, db) {
   ]);
 
   // create Dataframe from profile data
-  let profileDF = new DataIngestor(
-    profileData,
-    profileName,
-    isAggregate,
-  ).processRaw();
-
-  // join with previous profile data Dataframe
-  profileDF = profileDF.join(
-    new DataIngestor(
-      previousProfileData,
-      profileName,
-      isAggregate,
-      /* is previous */ true,
-    ).processRaw(),
-    'variable',
-  );
-
-  // join with compare profile data Dataframe
-  profileDF = profileDF.join(
-    // 'compare' profiles are always NOT aggregate; they are comprised of a single geoid
-    new DataIngestor(
-      compareProfileData,
-      profileName,
-      /* is aggregate */ false,
-      /* is previous */ false,
-      /* is comparison */ true,
-    ).processRaw(),
-    'variable',
-  );
-
-  // convert combined Dataframe to array of objects
-  // (change & difference calculations requires ad-hoc property addition,
-  // which is *MUCH* easier with objects than Dataframe rows)
-  // TODO: confirm using Dataframe is faster than just iterating over objects...
-  return profileDF
-    .toCollection()
+  new DataProcessor(profileData, profileName, isAggregate).process();
+  new DataProcessor(previousProfileData, profileName, isAggregate, /* isPrevious */ true).process();
+  new DataProcessor(compareProfileData, profileName, /* isAggregate */ false).process();
+
+  join(profileData, previousProfileData, compareProfileData);
+
+  
+  profileData
     .map((row) => {
       const varConfig = find(specialCalculationConfigs[profileName], ['variable', row.variable]) || {};
       doChangeCalculations(row, varConfig);
       doDifferenceCalculations(row);
-      removePercentsFromSpecialVars(row, varConfig);
-      return row;
-    });
+  });
+  return profileData;
+}
+
+/* 
+ * Joins profile, previousProfile, and compareProfile row objects,
+ * prepending key names with appropriate prefixes before combining.
+ * (previous and compare, respectively).
+ * @param{Object[]} profile - Array of profile row objects
+ * @param{Object[]} previous - Array of previous profile row objects
+ * @param{Object[]} compare - Array of compare profile row objects
+ */
+function join(profile, previous, compare) {
+  const valueKeys = getValueKeys(Object.keys(profile[0]));
+
+  profile.sort(profileSort);   
+  previous.sort(profileSort);
+  compare.sort(profileSort);
```

Comment:
I think reading this I just _trust_ that these get sorted by _something_ though it wasn't clear what was the sort key. `profileSort` is defined at the bottom so it can be a little mysterious what's happening. I don't think there's a way to make `profileSort` more configurable, though, without jumping through some weird hoops:

```javascript
profile.sort(sortBy('variable'))`

function sortBy(sortKey) {
  return function(rowA, rowB) {
    if (rowA[sortKey] > rowB[sortKey]) return 1;
    if (rowA[sortKey] < rowB[sortKey]) return -1;
    return 0;
  }
}
```

~Would it make sense to make `profileSort` more generic by allowing passage of the target column name (`variable`)?~

Reply:
perhaps best compromise here is to rename the sort function to something like `sortRowByVariable`